src/mlpack/core/data/scaler_methods/whitening.hpp

```diff
@@ -0,0 +1,147 @@
+ * @endcode
+ */
+class Whitening
+{
+ public:
+  /**
+  * A constructor to set the regulatization parameter.
+  *
+  * @param eps Regularization parameter.
+  */
+  Whitening(double eps = 0.00005)
+  {
+    epsilon = eps;
+  }
+
+  /**
+  * Function for PCA whitening.
+  *
+  * @param input Dataset to scale features.
+  * @param output Output matrix with whitened features.
+  */
+  template<typename MatType>
+  void PCA(const MatType& input, MatType& output)
+  {
+    output.copy_size(input);
+    itemMean = arma::mean(input, 1);
+    // Centering the Dataset
+    output = (input.each_col() - itemMean);
+    arma::mat diag;
+    // Get eigenvectors and eigenvalues of covariance of input matrix.
+    eig_sym(eigenValues, eigenVectors, mlpack::math::ColumnCovariance(output));
+    // Generate diagonal matrix using 1 / sqrt(eigenvalues + epsilon) for each 
+    // value.
+    for (size_t i = 0; i < eigenValues.n_elem; i++)
+        eigenValues(i) = eigenValues(i) + epsilon;
+    mlpack::math::VectorPower(eigenValues, -0.5);
+    diag.zeros(eigenValues.n_elem, eigenValues.n_elem);
+    diag.diag() = eigenValues;
+
+    // Our whitening matrix is diag(1 / sqrt(eigenvectors)) * eigenvalues.
+    whiteningMatrix = diag * trans(eigenVectors);
+
+    // Now apply the whitening matrix.
+    output = whiteningMatrix * output;
+  }
+
+  /**
+  * Function for ZCA whitening.
+  *
+  * @param input Dataset to scale features.
+  * @param output Output matrix with whitened features.
+  */
+  template<typename MatType>
+  void ZCA(const MatType& input, MatType& output)
+  {
+    output.copy_size(input);
+    itemMean = arma::mean(input, 1);
+    // Centering the Dataset
+    output = (input.each_col() - itemMean);
+    arma::mat diag;
+    // Get eigenvectors and eigenvalues of covariance of input matrix.
+    eig_sym(eigenValues, eigenVectors, mlpack::math::ColumnCovariance(output));
+    // Generate diagonal matrix using 1 / sqrt(eigenvalues + epsilon) for each 
+    // value.
+    for (size_t i = 0; i < eigenValues.n_elem; i++)
+        eigenValues(i) = eigenValues(i) + epsilon;
+    mlpack::math::VectorPower(eigenValues, -0.5);
+    diag.zeros(eigenValues.n_elem, eigenValues.n_elem);
+    diag.diag() = eigenValues;
+
+    // Our whitening matrix is diag(1 / sqrt(eigenvectors)) * eigenvalues.
+    whiteningMatrix = diag * trans(eigenVectors);
+
+    // Now apply the whitening matrix.
+    output = whiteningMatrix * output;
+    // ZCA output.
+    output = eigenVectors * output;
+  }
+
+  //! Get the Mean row vector.
+  const arma::vec& ItemMean() const { return itemMean; }
+  //! Get the eigenvalues vector.
+  const arma::vec& EigenValues() const { return eigenValues; }
+  //! Get the eigenvector.
+  const arma::mat& EigenVectors() const { return eigenVectors; }
+  //! Get the WhiteningMatrix.
+  const arma::mat& WhiteningMatrix() const { return whiteningMatrix; }
+  //! Get the Regularisation Parameter.
+  const double& Epsilon() const { return epsilon; }
+
```

Comment:
Should we also add an `InverseTransform()` function?

Reply:
I agree with you, Works smoothly. Sometimes my computer behaves in weird ways and after having discussion about the function , i tested it with some inputs and it works smoothly.The same code which i wrote previous worked well.